### ISSUE TITLE
core::0133::request-id-field false positive for AIP-155 idempotency keys

### ISSUE THREAD

**Issue by mo-cmyk** [googleapis/api-linter#1574]
## Summary

The `core::0133::request-id-field` rule flags Create requests as missing a client-specified ID field (e.g. `book_id`), but this creates false positives when `request_id` is used as an **idempotency key** per [AIP-155](https://google.aip.dev/155).

## Problem

AIP-133 allows client-specified resource IDs via fields like `{resource}_id` (e.g., `book_id` in `CreateBookRequest`).

AIP-155 recommends `request_id` as an idempotency key - a completely different use case:
- Purpose: deduplicate mutation requests from network retries
- Behavior: server returns cached response for duplicate request IDs
- Not a resource identifier

The linter strictly enforces AIP-133 and requires `{resource}_id`, failing to recognize that `request_id` satisfies the idempotency requirement.

## Example

```protobuf
message CreateBookRequest {
  string parent = 1 [(google.api.resource_reference).child_type = "library.googleapis.com/Book"];
  
  Book book = 2 [(google.api.field_behavior) = REQUIRED];
  
  // Idempotency key per AIP-155 - NOT a client-specified resource ID
  string request_id = 3 [
    (google.api.field_behavior) = OPTIONAL,
    (google.api.field_info).format = UUID4
  ];
}
```

This triggers:
```
core::0133::request-id-field: create methods should contain a singular `string book_id` field.
```

## Expected Behavior

The rule should not flag the missing `{resource}_id` when a valid AIP-155 idempotency key is present:
- Field name is exactly `request_id`
- Field behavior is `OPTIONAL`
- Field has `UUID4` format annotation

## Suggested Fix

Update the rule to check for `request_id` as a valid alternative when `{resource}_id` is missing.

## References

- [AIP-133: Create](https://google.aip.dev/133) - Client-specifiedgh issue edit 1574 --repo googleapis/api-linter --body-file - <<EOF
## Summary

The `core::0133::request-id-field` rule flags Create requests as missing a client-specified ID field (e.g. `book_id`), but this creates false positives when `request_id` is used as an **idempotency key** per [AIP-155](https://google.aip.dev/155).

## Problem

AIP-133 allows client-specified resource IDs via fields like `{resource}_id` (e.g., `book_id` in `CreateBookRequest`).

AIP-155 recommends `request_id` as an idempotency key - a completely different use case:
- Purpose: deduplicate mutation requests from network retries
- Behavior: server returns cached response for duplicate request IDs
- Not a resource identifier

The linter strictly enforces AIP-133 and requires `{resource}_id`, failing to recognize that `request_id` satisfies the idempotency requirement.

## Example

```protobuf
message CreateBookRequest {
  string parent = 1 [(google.api.resource_reference).child_type = "library.googleapis.com/Book"];
  
  Book book = 2 [(google.api.field_behavior) = REQUIRED];
  
  // Idempotency key per AIP-155 - NOT a client-specified resource ID
  string request_id = 3 [
    (google.api.field_behavior) = OPTIONAL,
    (google.api.field_info).format = UUID4
  ];
}
```

This triggers:
```
core::0133::request-id-field: create methods should contain a singular `string book_id` field.
```

## Expected Behavior

The rule should not flag the missing `{resource}_id` when a valid AIP-155 idempotency key is present:
- Field name is exactly `request_id`
- Field behavior is `OPTIONAL`
- Field has `UUID4` format annotation

## Suggested Fix

Update the rule to check for `request_id` as a valid alternative when `{resource}_id` is missing.

## References

- [AIP-133: Create](https://google.aip.dev/133) - Client-specified IDs
- [AIP-155: Request identification](https://google.aip.dev/155) - Idempotency keys
- [Rule documentation](https://linter.aip.dev/133/request-id-field)

**Comment by quirogas** [comment]
Hi @mo-cmyk, thank you for raising this issue. There is a distinction between the purposes of the two AIPs:

* **AIP-133** checks for user-specified ID requirements (e.g., `string {resource}_id`) to handle **Resource Identity**.
* **AIP-155** uses `string request_id` strictly for **Request Idempotency** (deduplication/retries).

AIP-155 does not exempt the requirement for `string {resource}_id`; they serve distinct purposes, and a standard Create method often requires both. Additionally, it was [previously decided](https://github.com/googleapis/api-linter/pull/1155) that the rule should apply to all Standard Create methods, so it shouldn't be restricted to declarative methods only.

However, you might have spotted a valid bug in how the linter enforces this. While `request_id` is not a substitute for `{resource}_id`, the linter should ensure it is flagging the missing field correctly rather than conflating the two. Additionally, the [rule docs](https://github.com/googleapis/api-linter/blob/1760e4935162cb59436e843109f3a19009349379/docs/rules/0133/request-id-field.md?plain=1#L13) still contain the "declarative-friendly" qualifier, which seems to have been missed when the logic was updated in PR #1155. We should definitely update the documentation to drop that phrase to avoid further confusion.

**Comment by mo-cmyk** [comment]
Thanks for the clarification @quirogas! That makes perfect sense now.

I understand:
- `{resource}_id` (AIP-133) = Resource Identity - required for Create methods
- `request_id` (AIP-155) = Idempotency key - separate purpose, not a substitute

I've updated the PR to align with this understanding:

1. **Reverted the code** - the rule now applies to all Create methods (as intended after PR #1155)
2. **Updated documentation** - removed the 'declarative-friendly' qualifier that was outdated
3. **Reverted tests** - back to original state

The PR is now a documentation-only fix to remove the misleading 'declarative-friendly' wording from the rule docs.

Regarding my original issue - I now understand that the linter is behaving correctly. If my API needs client-specified IDs, I should add `{resource}_id`. If I don't need that behavior, I can disable the rule with the appropriate comment.

**Comment by quirogas** [comment]
Thank you @mo-cmyk. For posterity, a few exceptions exist in [API-133](https://google.aip.dev/133) for User-specified IDs, which are also documented under [User-specified IDs](https://google.aip.dev/133#user-specified-ids):

> On the [data plane](https://google.aip.dev/111#data-plane), an API should allow a user to specify the ID. Exceptional cases should have the following behavior:
> 
> The data plane resource allows identical records without a need to disambiguate between the two (e.g. rows in a table with no primary key).
> The data plane resource will not be exposed in [Declarative clients](https://google.aip.dev/9#declarative-clients).

These exceptions might actually conflict with the presence of `request_id`, so we should verify that the linter does not flag these cases. If the `api-linter` does generate some findings for this cases,  we can  add request_id to an allowlist to prevent collisions with the `request_id` field.

**Comment by mo-cmyk** [comment]
Thanks for the additional context @quirogas! That's really helpful.

I see - so there are valid exceptions in AIP-133 where `{resource}_id` is not required:
- Data plane resources allowing identical records (no primary key)
- Data plane resources not exposed in Declarative clients

And in these cases, having `request_id` without `{resource}_id` would be legitimate, but the linter might still flag it.

**Question**: Should the `request_id` allowlist enhancement be tracked as a separate issue/PR? I'd like to keep the current PR focused on the documentation fix, and we can address the allowlist logic separately if that works for you.

Happy to create a follow-up issue for the allowlist enhancement if you'd like!

**Comment by mo-cmyk** [comment]
**Update**: I tested the scenario you mentioned and confirmed the linter behaves correctly.

### Test Result
Added a test case `InvalidRequestIdOnly` - when a Create request has `request_id` but no `book_id`:
- ✅ **Linter correctly flags** the missing `book_id`
- `request_id` is not treated as a substitute for `{resource}_id`

```go
// request_id (AIP-155) is NOT a substitute for {resource}_id (AIP-133)
{"InvalidRequestIdOnly", "string request_id = 2;", problems},
```

### PR Update
- Rebased on latest `main`
- Added the test case to document this behavior
- PR is ready for merge (already approved)

Let me know if you'd like me to open a separate issue for the allowlist enhancement for the exception cases you mentioned!

**Comment by quirogas** [comment]
@mo-cmyk Thank you! I went ahead and merged the PR. I do think it would be worth opening a follow-up PR to allowlist `request_id`, referencing this same issue.

**Comment by mo-cmyk** [comment]
Created follow-up PR #1583 to implement the `request_id` allowlist as suggested.

The PR:
- Skips the `{resource}_id` check if `request_id` field is present
- Adds test cases for the allowlist behavior
- Updates documentation with a note about the AIP-133 exception cases

Thanks for the guidance @quirogas!

**Comment by quirogas** [comment]
Hi @mo-cmyk, continuing from PR #1583.

I would like to apologize for the misunderstanding and for perhaps making things a bit confusing. I should clarify that AIP-133 and AIP-155 **are not mutually exclusive**, and therefore, a clear implementation of both is more complex than I might have communicated.

AIP-155 (`request_id`) handles **network reliability** (idempotency), while AIP-133 (`{resource}_id`) handles **resource identity** (naming). Implementing one does not satisfy the requirement for the other.

**Why the allowlist is unsafe:**
Most standard APIs (Control Plane) follow **Example 1** below. If we allowed `request_id` to bypass the check for `{resource}_id`, the linter would fail to flag missing ID fields in standard Create methods.

**1. Standard Control Plane Create Request (AIP-133)**
*Requires both distinct fields.*
```protobuf
message CreateBookRequest {
  // ...
  Book book = 2 [(google.api.field_behavior) = REQUIRED];

  // User-assigned ID (AIP-133)
  string book_id = 3 [(google.api.field_behavior) = OPTIONAL];

  // Idempotency Key (AIP-155)
  string request_id = 4 [(google.api.field_behavior) = OPTIONAL];
}
```

**2. Standard Create Request without Idempotency**
*Requires resource ID, but idempotency is optional.*
It is perfectly valid to have a standard create method that requires a user-specified ID but does not support an idempotency key.
```protobuf
message CreateBookRequest {
  // ...
  Book book = 2 [(google.api.field_behavior) = REQUIRED];

  // User-assigned ID (AIP-133) is REQUIRED for compliance here.
  string book_id = 3 [(google.api.field_behavior) = OPTIONAL];

  // request_id is NOT present.
}
```

**3. Data Plane Create Request - No `request_id` (AIP-155)**
*Server-assigned ID, no idempotency.*
```protobuf
message CreateBookContentRequest {
  // ...
  Book book = 2 [(google.api.field_behavior) = REQUIRED];
  // No book_id, No request_id
}
```

**4. Data Plane Create Request - With `request_id` (AIP-155)**
*Server-assigned ID, but with idempotency.*
```protobuf
message CreateBookContentRequest {
  // ...
  Book book = 2 [(google.api.field_behavior) = REQUIRED];

  // Idempotency Key (AIP-155) is present, but book_id is still rightly omitted.
  string request_id = 3 [(google.api.field_behavior) = OPTIONAL];
}
```

If your API falls into the category of **Example 4** (Data Plane/Server-assigned ID), this is considered an exception to the rule. Since the linter cannot distinguish Example 4 from a malformed Example 1, we ask that you disable the rule explicitly for that message:

```protobuf
// api-linter: core::0133::request-id-field=disabled
```

Thanks for working through this, and I hope this clears any misunderstanding.
